src/libraries/Microsoft.PowerFx.Core/Public/DisplayNames/DisplayNameProvider.cs

```diff
@@ -0,0 +1,52 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Text;
+using System.Linq;
+using Microsoft.PowerFx.Core.UtilityDataStructures;
+
+namespace Microsoft.PowerFx.Core
+{
+    internal class DisplayNameProvider
+    {
+        // First is Logical Name, Second is Display Name
+        private BidirectionalDictionary<string, string> _displayNames;
+
+        public DisplayNameProvider()
+        {
+            _displayNames = new(StringComparer.Ordinal, StringComparer.Ordinal);
+        }
+
+        public bool TryAddField(string logicalName, string displayName)
+        {
+            // Check for collisions between display and logical names
+            if (_displayNames.ContainsSecondKey(logicalName) || _displayNames.ContainsFirstKey(logicalName) ||
+                _displayNames.ContainsFirstKey(displayName) || _displayNames.ContainsSecondKey(displayName))
+            {
+                return false; 
+            }
+            return _displayNames.Add(logicalName, displayName);
+        }
+
+
+        public bool TryGetLogicalName(string displayName, out string logicalName)
+        {
+            return _displayNames.TryGetFromSecond(displayName, out logicalName);
+        }
+        
+        public bool TryGetDisplayName(string logicalName, out string displayName)
+        {
+            return _displayNames.TryGetFromFirst(logicalName, out displayName);
+        }
+
+        public bool Matches(DisplayNameProvider other)
```

Comment:
Is this O[N^2]?

Reply:
No, it's O(N). Iterates each set of keys, and checks that it's a member of the other set, which is backed by a hash table.

Reply:
per other conversations, we may just want to avoid calling Matches() altogether. It's not an efficient impl, and it's dubious how a caller would invoke it 100% correctly.

Reply:
Ah, good point. Reference equality should be fine here. thanks.